### DISCUSSION TITLE
WhenAnyValue - how to ensure the order of async subscriber?

### DISCUSSION THREAD

**Discussion by sinatrocious** [reactiveui/ReactiveUI#3710] [Q&A] [answered]
I've following code:

```c#
SomeReactive.WhenAnyValue(o => o.SomeBool).Subscribe(async someBool =>
{
    if(someBool)
    {
        await ...
        Start();
    }
    else
    {
        await ...
        Stop();
    }
});
```

If `SomeBool` changes rapidly (and in different threads), then it may lead to a wrong order of calls:

```
Start()
Stop()
Stop()
Start()
```

Is there a way to "prevent next call of subscriber, until previous is finished"? I can use `lock` inside subscriber, but I was hoping for some kind of synchronization from reactive UI side. Is there anything I miss? Or perhaps I need to use another approach?

P.S.: I [asked this question](https://stackoverflow.com/q/77489039/1997232) on StackOverflow earlier.

**Comment by ChrisPulman** [comment]
If I was trying to do this I would use `Synchronize` as follows:
```
var locker = new object();
SomeReactive.WhenAnyValue(o => o.SomeBool)
	.Synchronize(locker)
	.Subscribe(async someBool =>
	{
	    if(someBool)
	    {
	        await ...
	        Start();
	    }
	    else
	    {
	        await ...
	        Stop();
	    }
	});
```
Hopefully this works in your situation.

**Reply by glennawatson** [reply]
Potentially you could use ObserveOn with a thread scheduler

**Reply by sinatrocious** [reply]
@glennawatson, I doubt it. Maybe I don't understand something fully, but `await` inside subscriber is my problem: thread is released, so scheduler will continue processing next change (talking about `CurrentThreadScheduler`). Unless you mean writing some custom scheduler, which is capable of figuring out that subsciber is `async` and `await` for it?

**Reply by anaisbetts** [reply]
No, the solution to this problem is to get your source of asynchrony (the await function) inside the Pipeline - when you write your async method inside the Subscribe, that's a void async, it's completely outside of Rx. Instead, you need to put your async functions inside a SelectMany or a Select.Defer.Concat and allow Rx to constrain the asynchrony. That way it handles all the locking and reentrancy (what you're experiencing) for you

**Reply by sinatrocious** [reply, marked as answer]
@anaisbetts, `async void` is the problem indeed. I've seen [your answer](https://stackoverflow.com/a/23018227/1997232) in SO topic, where Stephen Cleary explains it ;)

I will use `SubscribeAsync` in place of `Subscribe`:

```c#
public static IDisposable SubscribeAsync<T>(this IObservable<T> source, Func<T, Task> func) =>
            source.Select(o => Observable.FromAsync(_ => func(o)))
                .Concat()
                .Subscribe();
```
